sklearn/manifold/_utils.pyx

```diff
@@ -12,24 +12,20 @@ cdef float PERPLEXITY_TOLERANCE = 1e-5
 
 @cython.boundscheck(False)
 cpdef np.ndarray[np.float32_t, ndim=2] _binary_search_perplexity(
-        np.ndarray[np.float32_t, ndim=2] affinities,
-        np.ndarray[np.int64_t, ndim=2] neighbors,
+        np.ndarray[np.float32_t, ndim=2] distances,
         float desired_perplexity,
         int verbose):
     """Binary search for sigmas of conditional Gaussians.
 
     This approximation reduces the computational complexity from O(N^2) to
-    O(uN). See the exact method '_binary_search_perplexity' for more details.
+    O(uN).
 
     Parameters
     ----------
-    affinities : array-like, shape (n_samples, k)
-        Distances between training samples and its k nearest neighbors.
-
-    neighbors : array-like, shape (n_samples, k) or None
-        Each row contains the indices to the k nearest neigbors. If this
-        array is None, then the perplexity is estimated over all data
-        not just the nearest neighbors.
+    distances : array-like, shape (n_samples, n_neighbors)
+        Distances between training samples and their k nearest neighbors.
```

Comment:
Are these not ordinarily squared distances?

Reply:
it might then be better unchanged as affinity

Reply:
I would say 'distances' is a better name since a distance is smaller for closer points, whereas an affinity is larger for closer points.

The default is `TSNE(metric='euclidean')` which under the hood get transformed into squared euclidean, but TSNE accepts any kind of distances, even though the original derivation used only squared euclidean distance.

Reply:
dissimilarity?

Reply:
The function makes sense algorithmically under the assumption that the input are squared distances, not distances, even if we currently handle this weirdly in the non-euclidean metric processing. IMO we *should* be squaring all the distances.

`sqdistances` would make me happier, I think, in terms of keeping this maintainable.